### PR TITLE
Plugin uninstall - public API

### PR DESCRIPTION
**Pull request by lpenet**
Companion PR of https://github.com/dataiku/dip/pull/10249

### CODE REVIEW

File: dataikuapi/dss/plugin.py [hunk 1]
```diff
@@ -116,6 +116,73 @@ def update_from_git(self, repository_url, checkout = "master", subpath=None):
...
+        Each element of the usages list contains:
+        - an elementKind of the plugin element, such as webapps, python-probes, python-checks, etc.
+        - an elementType of the plugin element,
+        - the objectType and objectId of the object using this plugin element, along with their projectKey,
+        if pertinent. Some objects, for instance clusters, are not contained in a project.
+
+        Some custom types may not be found during the analysis. This typically occurs when a plugin was removed,
+        while still being used. This prevents further analysis of the object relying on this type and may hide
+        some uses of the plugin. Thus, those missingTypes are enumerated in the missingTypes list, which
+        includes the missingType along with the same information on the object as for usages.
```
[the remark this answers is not in the document]
Comment (lpenet) [reply dataikuapi/dss/plugin.py:140]: But informative, no ?

[the remark this answers is not in the document]
Comment (instanceofme) [reply dataikuapi/dss/plugin.py:140]: Yes, I apparently didn't save my suggestion 😅 

```suggestion
        Some custom types may not be found during usage analysis, typically when a plugin was removed
        but is still used. This prevents some detailed analysis and may hide some uses.
        This information is provided in missingTypes.
```

File: dataikuapi/dss/plugin.py [hunk 2]
```diff
@@ -116,6 +116,73 @@ def update_from_git(self, repository_url, checkout = "master", subpath=None):
...
+    ########################################################
+    # Plugin uninstall/delete
+    ########################################################
+
+    def list_usages(self, project_key=None):
+        """
+        Get the list of usages of the plugin.
+
+        Returns a dict with two keys:
+        - usages, a list of (elementKind, elementType, objectType, projectKey, objectId) tuples
```
[the remark this answers is not in the document]
Comment (lpenet) [reply dataikuapi/dss/plugin.py:128]: In fact, it was not a tuple (my bad) but a dict. I coded classes, anyway.

File: dataikuapi/dss/plugin.py [hunk 3]
```diff
@@ -116,6 +116,73 @@ def update_from_git(self, repository_url, checkout = "master", subpath=None):
...
+        - missingTypes, a list of (missingType, objectType, projectKey, objectId) tuples
+
+        Each element of the usages list contains:
+        - an elementKind of the plugin element, such as webapps, python-probes, python-checks, etc.
+        - an elementType of the plugin element,
+        - the objectType and objectId of the object using this plugin element, along with their projectKey,
+        if pertinent. Some objects, for instance clusters, are not contained in a project.
```
[the remark this answers is not in the document]
Comment (lpenet) [reply dataikuapi/dss/plugin.py:135]: ok

File: dataikuapi/dss/plugin.py [hunk 4]
```diff
@@ -116,6 +116,73 @@ def update_from_git(self, repository_url, checkout = "master", subpath=None):
...
+
+        Some custom types may not be found during the analysis. This typically occurs when a plugin was removed,
+        while still being used. This prevents further analysis of the object relying on this type and may hide
+        some uses of the plugin. Thus, those missingTypes are enumerated in the missingTypes list, which
+        includes the missingType along with the same information on the object as for usages.
+
+        :param str project_key: optional key of project where to look for usages. Default is None and looking in all projects.
+        :return: dict
```
[the remark this answers is not in the document]
Comment (lpenet) [reply dataikuapi/dss/plugin.py:143]: ok

File: dataikuapi/dss/plugin.py [hunk 5]
```diff
@@ -116,6 +116,73 @@ def update_from_git(self, repository_url, checkout = "master", subpath=None):
...
+        """
+        Request pre-deletion checks, as aggregated information on the usage of the plugin.
+
+        Information is provided as a dict with the following entries:
+        - projectCount: count of projects using at least an element of this plugin
+        - usedElemCount: count of elements of this plugin in use
+        - objectAnalysisErrors: count of errors encountered while analyzing usages.
+
+        Detailed information can be obtained by calling :func:`list_usages`.
+        :return: dict
```
[the remark this answers is not in the document]
Comment (lpenet) [reply dataikuapi/dss/plugin.py:160]: ok.

File: dataikuapi/dss/plugin.py [hunk 6]
```diff
@@ -116,6 +116,73 @@ def update_from_git(self, repository_url, checkout = "master", subpath=None):
...
+    # Plugin uninstall/delete
+    ########################################################
+
+    def list_usages(self, project_key=None):
+        """
+        Get the list of usages of the plugin.
+
+        Returns a dict with two keys:
+        - usages, a list of (elementKind, elementType, objectType, projectKey, objectId) tuples
+        - missingTypes, a list of (missingType, objectType, projectKey, objectId) tuples
```
[the remark this answers is not in the document]
Comment (lpenet) [reply dataikuapi/dss/plugin.py:129]: ok.

File: dataikuapi/dss/plugin.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lpenet) [reply dataikuapi/dss/plugin.py:140]: ok.

File: dataikuapi/dss/plugin.py [hunk 7]
```diff
@@ -147,4 +175,148 @@ def put_file(self, path, f):
...
 
-        
\ No newline at end of file
+
+class DSSPluginUsage(object):
+    """
+    Information on a usage of an element of a plugin.
+    """
+    def __init__(self, json_object):
+        self._json_object = json_object
```
Comment (instanceofme) [inline dataikuapi/dss/plugin.py:184]: [Nitpick] I would rename the field as `_dict` or `_data` as it is not JSON anymore. Same for argument.

File: dataikuapi/dss/plugin.py [hunk 8]
```diff
@@ -147,4 +175,148 @@ def put_file(self, path, f):
...
+        """
+        return self._json_object.get("projectKey", None)
+
+
+class DSSMissingType(object):
+    """
+    Information on a type not found while analyzing usages of a plugin.
+    """
+    def __init__(self, json_object):
+        self._json_object = json_object
```
Comment (instanceofme) [inline dataikuapi/dss/plugin.py:237]: Ditto

File: dataikuapi/dss/plugin.py [hunk 9]
```diff
@@ -147,4 +175,148 @@ def put_file(self, path, f):
...
+
+    def get_raw(self):
+        return self._json_object
```
Comment (instanceofme) [inline dataikuapi/dss/plugin.py:240]: Can add docstring

File: dataikuapi/dss/plugin.py [hunk 10]
```diff
@@ -147,4 +175,148 @@ def put_file(self, path, f):
...
+        """
+        :return: the missing type
+        :rtype: str
+        """
+        return self._json_object["missingType"]
+
+    @property
+    def object_id(self):
+        """
+        :return: Id of the object relying on the missing type
```
Comment (instanceofme) [inline dataikuapi/dss/plugin.py:253]: Not sure I understand "relying on"?

File: dataikuapi/dss/plugin.py [hunk 11]
```diff
@@ -147,4 +175,148 @@ def put_file(self, path, f):
...
+        :rtype: str or none
+        """
+        return self._json_object.get("projectKey", None)
+
+
+class DSSPluginUsages(object):
+    """
+    Information on the usages of a plugin.
+
+    Contains both usages (a list of instances of :class:`DSSPluginUsage`) and analysis errors, if any
```
Comment (instanceofme) [inline dataikuapi/dss/plugin.py:279]:
```suggestion
    Contains both usages (a list of :class:`DSSPluginUsage`) and analysis errors, if any
```

File: dataikuapi/dss/plugin.py [hunk 12]
```diff
@@ -147,4 +175,148 @@ def put_file(self, path, f):
...
+    (a list of instances of :class:`DSSMissingType`).
```
Comment (instanceofme) [inline dataikuapi/dss/plugin.py:280]:
```suggestion
    (a list of :class:`DSSMissingType`).
```

File: dataikuapi/dss/plugin.py [hunk 13]
```diff
@@ -147,4 +175,148 @@ def put_file(self, path, f):
...
+    but is still used. This prevents some detailed analysis and may hide some uses.
+    This information is provided in missingTypes.
+    """
+    def __init__(self, json_object):
+        """
+        :param dict json_object: the usages as json dict
+        :param list(:class:`DSSPluginUsage`) usages: plugin usages
+        :param list(:class:`DSSMissingType`) missing_types:
+        """
+        self._json_object = json_object
```
Comment (instanceofme) [inline dataikuapi/dss/plugin.py:292]: Rename

File: dataikuapi/dss/plugin.py [hunk 14]
```diff
@@ -147,4 +175,148 @@ def put_file(self, path, f):
...
+        self._usages = []
+        self._missing_types = []
+        for json_usage in json_object.get("usages", []):
+            self._usages.append(DSSPluginUsage(json_usage))
+        for json_missing_type in json_object.get("missingTypes"):
+            self._missing_types.append(DSSMissingType(json_missing_type))
+
+    def get_raw(self):
+        return self._json_object
```
Comment (instanceofme) [inline dataikuapi/dss/plugin.py:301]: Docstring

File: dataikuapi/dss/plugin.py [hunk 15]
```diff
@@ -147,4 +175,148 @@ def put_file(self, path, f):
...
+
+    def needs_force_delete(self):
+        return not (not self._usages and not self._missing_types)
```
Comment (instanceofme) [inline dataikuapi/dss/plugin.py:304]: Docstring

**Comment by lpenet** [comment]
> `@property` is overkill IMO and we could have simple fields, and just use the class docstring to document what's in them…

I do not think so. If we do the way you suggest, we shall have no completion by the editor.

File: dataikuapi/dss/plugin.py (hunk 7 above)
Comment (lpenet) [reply dataikuapi/dss/plugin.py:184]: ok.

File: dataikuapi/dss/plugin.py (hunk 9 above)
Comment (lpenet) [reply dataikuapi/dss/plugin.py:240]: ok.

File: dataikuapi/dss/plugin.py (hunk 10 above)
Comment (lpenet) [reply dataikuapi/dss/plugin.py:253]: Replaced by "depending". Hope it is clearer.

File: dataikuapi/dss/plugin.py (hunk 11 above)
Comment (lpenet) [reply dataikuapi/dss/plugin.py:279]: ok.

File: dataikuapi/dss/plugin.py (hunk 12 above)
Comment (lpenet) [reply dataikuapi/dss/plugin.py:280]: ok.

File: dataikuapi/dss/plugin.py (hunk 13 above)
Comment (lpenet) [reply dataikuapi/dss/plugin.py:292]: ok.

File: dataikuapi/dss/plugin.py (hunk 14 above)
Comment (lpenet) [reply dataikuapi/dss/plugin.py:301]: ok.

File: dataikuapi/dss/plugin.py (hunk 15 above)
Comment (lpenet) [reply dataikuapi/dss/plugin.py:304]: ok.

File: dataikuapi/dss/plugin.py (hunk 8 above)
Comment (lpenet) [reply dataikuapi/dss/plugin.py:237]: ok.

File: dataikuapi/dss/plugin.py [hunk 16]
```diff
@@ -147,4 +175,148 @@ def put_file(self, path, f):
...
+        return self._json_object["elementKind"]
+
+    @property
+    def element_type(self):
+        """
+        Element type
+        :return: the element type
+        :rtype: str
+        """
+        return self._json_object["elementType"]
```
Comment (NicolasTriquenaux) [inline dataikuapi/dss/plugin.py]: Here and above you do not use `.get()`. So you expect these field to always exist ? If not you you might want to catch the KeyError exception that can be raised and re-raise it with a more comprehensive message for the API user ?

File: dataikuapi/dss/plugin.py [hunk 17]
```diff
@@ -147,4 +175,95 @@ def put_file(self, path, f):
...
+    """
+    def __init__(self, element_kind, element_type, object_id, object_type, project_key):
+        """
+
+        :param str element_kind:
+        :param str element_type:
+        :param str object_id:
+        :param str object_type:
+        :param str project_key:
+        """
```
[the remark this answers is not in the document]
Comment (lpenet) [reply dataikuapi/dss/plugin.py]: Outdated.

File: dataikuapi/dss/plugin.py [hunk 18]
```diff
@@ -116,6 +107,43 @@ def update_from_git(self, repository_url, checkout = "master", subpath=None):
...
+        :return: a :class:`dataikuapi.dssfuture.DSSFuture`
+        """
+
+        params = {
+            "force": force
+        }
+        ret = self.client._perform_json("POST", "/plugins/{pluginId}/actions/delete".format(pluginId=self.plugin_id),
+                                        body=params)
+        if "projectCount" in ret:
+            raise Exception("Plugin has usages or analysis errors.")
```
[the remark this answers is not in the document]
Comment (lpenet) [reply dataikuapi/dss/plugin.py:145]: We could raise a `DataikuException`. But I see not case of a more specific exception in dataiku-api-client-python...

WDYT ?

[the remark this answers is not in the document]
Comment (lpenet) [reply dataikuapi/dss/plugin.py:145]: So, we discussed and go for `DataikuException`.

**Comment by NicolasTriquenaux** [comment]
> `@property` is overkill IMO and we could have simple fields, and just use the class docstring to document what's in them…

In tha specific case it can be overkill. But it is also a sign that if you want to alter any member you should use the property setter and not access and edit directly the class member.

**Comment by instanceofme** [comment]
@NicolasTriquenaux not sure why the PR was closed?

**Comment by NicolasTriquenaux** [comment]
> @NicolasTriquenaux not sure why the PR was closed?

Wrong manip sorry.

File: dataikuapi/dss/plugin.py [hunk 19]
```diff
@@ -147,4 +179,89 @@ def put_file(self, path, f):
...
+class DSSMissingType(object):
+    """
+    Information on a type not found while analyzing usages of a plugin.
+
+    Has the following properties:
+    - missing type: the missing type
+    - object_id: id of the object depending on the missing type. Can be None.
+    - object_type: type of the object depending on the missing type. Can be None.
+    - project_key: project key of the object depending on the missing type. Can be None.
+
```
Comment (instanceofme) [inline dataikuapi/dss/plugin.py:215]:
```suggestion
```

File: dataikuapi/dss/plugin.py [hunk 20]
```diff
@@ -116,6 +111,45 @@ def update_from_git(self, repository_url, checkout = "master", subpath=None):
...
+        params = {
+            "force": force
+        }
+        ret = self.client._perform_json("POST", "/plugins/{pluginId}/actions/delete".format(pluginId=self.plugin_id),
+                                        body=params)
+        # we check for the presence of a field of the aggregated usage statistics structure returned
+        # when the plugin exist and has usages
+        if "projectCount" in ret:
+            raise DataikuException("Plugin has usages or analysis errors. See list_usages() for more info.")
```
Comment (instanceofme) [inline dataikuapi/dss/plugin.py:150]: This actually never happens, as we are always in one of 3 cases:
1. Delete an unused plugin: works + `ret` is a future result which does not have `projectCount`
2. Delete a used plugin with `force=True`: same
3. Delete a used plugin without `force=True`: exception is raised server-side, `_perform_json` throws

The  behavior is ok but those lines are therefore unused / a bit misleading, seems to me that we can simply remove them.

```suggestion
```

**Pull request merged**